compile error OpenMP

Compile error here.
Using g++ 9.1.0 on Arch Linux

Commit 84175ba7c8c73dd4028f217daf7a5f4c25c2adaf

```
In file included from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/methods.hpp:17,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/embed.hpp:11,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/tapkee.hpp:10,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/tapkee_shogun.cpp:20:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp: In function ‘tapkee::SparseWeightMatrix tapkee::tapkee_internal::linear_weight_matrix(const RandomAccessIterator&, const RandomAccessIterator&, const Neighbors&, PairwiseCallback, tapkee::ScalarType, tapkee::ScalarType) [with RandomAccessIterator = __gnu_cxx::__normal_iterator<int*, std::vector<int> >; PairwiseCallback = pimpl_kernel_callback<shogun::Kernel>]’:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:108:15: error: ‘k’ not specified in enclosing ‘parallel’
  108 |   DenseMatrix gram_matrix = DenseMatrix::Zero(k,k);
      |               ^~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:105:9: error: enclosing ‘parallel’
  105 | #pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:133:38: error: ‘trace_shift’ not specified in enclosing ‘parallel’
  133 |    gram_matrix.diagonal().array() += trace_shift*trace;
      |                                      ^~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:105:9: error: enclosing ‘parallel’
  105 | #pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:137:61: error: ‘shift’ not specified in enclosing ‘parallel’
  137 |    SparseTriplet diagonal_triplet(index_iter,index_iter,1.0+shift);
      |                                                             ^~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:105:9: error: enclosing ‘parallel’
  105 | #pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp: In function ‘tapkee::SparseWeightMatrix tapkee::tapkee_internal::tangent_weight_matrix(RandomAccessIterator, RandomAccessIterator, const Neighbors&, PairwiseCallback, tapkee::IndexType, tapkee::ScalarType) [with RandomAccessIterator = __gnu_cxx::__normal_iterator<int*, std::vector<int> >; PairwiseCallback = pimpl_kernel_callback<shogun::Kernel>]’:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:38:15: error: ‘k’ not specified in enclosing ‘parallel’
   38 |   DenseMatrix gram_matrix = DenseMatrix::Zero(k,k);
      |               ^~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:35:9: error: enclosing ‘parallel’
   35 | #pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:40:39: error: ‘target_dimension’ not specified in enclosing ‘parallel’
   40 |   DenseMatrix G = DenseMatrix::Zero(k,target_dimension+1);
      |                                       ^~~~~~~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:35:9: error: enclosing ‘parallel’
   35 | #pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:69:18: error: ‘shift’ not specified in enclosing ‘parallel’
   69 |    SparseTriplet diagonal_triplet(index_iter,index_iter,shift);
      |                  ^~~~~~~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:35:9: error: enclosing ‘parallel’
   35 | #pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
      |         ^~~
In file included from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/methods.hpp:21,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/embed.hpp:11,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/tapkee.hpp:10,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/tapkee_shogun.cpp:20:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/diffusion_maps.hpp: In function ‘tapkee::DenseSymmetricMatrix tapkee::tapkee_internal::compute_diffusion_matrix(RandomAccessIterator, RandomAccessIterator, DistanceCallback, tapkee::ScalarType) [with RandomAccessIterator = __gnu_cxx::__normal_iterator<int*, std::vector<int> >; DistanceCallback = pimpl_distance_callback<shogun::Distance>]’:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/diffusion_maps.hpp:51:9: error: ‘n_vectors’ not specified in enclosing ‘parallel’
   51 | #pragma omp for nowait
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/diffusion_maps.hpp:48:9: error: enclosing ‘parallel’
   48 | #pragma omp parallel shared(diffusion_matrix,begin,callback) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/diffusion_maps.hpp:57:32: error: ‘width’ not specified in enclosing ‘parallel’
   57 |     ScalarType gk = exp(-(k*k)/width);
      |                                ^~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/diffusion_maps.hpp:48:9: error: enclosing ‘parallel’
   48 | #pragma omp parallel shared(diffusion_matrix,begin,callback) default(none)
      |         ^~~
In file included from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/methods.hpp:20,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/embed.hpp:11,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/tapkee.hpp:10,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/tapkee_shogun.cpp:20:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp: In function ‘tapkee::DenseSymmetricMatrix tapkee::tapkee_internal::compute_distance_matrix(RandomAccessIterator, RandomAccessIterator, PairwiseCallback) [with RandomAccessIterator = __gnu_cxx::__normal_iterator<int*, std::vector<int> >; PairwiseCallback = pimpl_distance_callback<shogun::Distance>]’:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp:122:9: error: ‘n_vectors’ not specified in enclosing ‘parallel’
  122 | #pragma omp for nowait
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp:119:9: error: enclosing ‘parallel’
  119 | #pragma omp parallel shared(begin,distance_matrix,callback) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp: In function ‘tapkee::DenseSymmetricMatrix tapkee::tapkee_internal::compute_distance_matrix(RandomAccessIterator, RandomAccessIterator, tapkee::tapkee_internal::Landmarks&, PairwiseCallback) [with RandomAccessIterator = __gnu_cxx::__normal_iterator<int*, std::vector<int> >; PairwiseCallback = pimpl_distance_callback<shogun::Distance>]’:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp:43:9: error: ‘n_landmarks’ not specified in enclosing ‘parallel’
   43 | #pragma omp for nowait
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp:40:9: error: enclosing ‘parallel’
   40 | #pragma omp parallel shared(begin,landmarks,distance_matrix,callback) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp: In function ‘tapkee::DenseMatrix tapkee::tapkee_internal::triangulate(RandomAccessIterator, RandomAccessIterator, PairwiseCallback, tapkee::tapkee_internal::Landmarks&, tapkee::DenseVector&, tapkee::tapkee_internal::EigendecompositionResult&, tapkee::IndexType) [with RandomAccessIterator = __gnu_cxx::__normal_iterator<int*, std::vector<int> >; PairwiseCallback = pimpl_distance_callback<shogun::Distance>]’:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp:85:15: error: ‘n_landmarks’ not specified in enclosing ‘parallel’
   85 |   DenseVector distances_to_landmarks(n_landmarks);
      |               ^~~~~~~~~~~~~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp:82:9: error: enclosing ‘parallel’
   82 | #pragma omp parallel shared(begin,end,to_process,distance_callback,landmarks, \
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp:87:9: error: ‘n_vectors’ not specified in enclosing ‘parallel’
   87 | #pragma omp for nowait
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp:82:9: error: enclosing ‘parallel’
   82 | #pragma omp parallel shared(begin,end,to_process,distance_callback,landmarks, \
      |         ^~~
In file included from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/methods.hpp:23,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/embed.hpp:11,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/tapkee.hpp:10,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/tapkee_shogun.cpp:20:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp: In function ‘tapkee::DenseSymmetricMatrix tapkee::tapkee_internal::compute_shortest_distances_matrix(RandomAccessIterator, RandomAccessIterator, tapkee::tapkee_internal::Neighbors&, DistanceCallback) [with RandomAccessIterator = __gnu_cxx::__normal_iterator<int*, std::vector<int> >; DistanceCallback = pimpl_distance_callback<shogun::Distance>]’:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp:55:13: error: ‘N’ not specified in enclosing ‘parallel’
   55 |   bool* f = new bool[N];
      |             ^~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp:53:9: error: enclosing ‘parallel’
   53 | #pragma omp parallel shared(shortest_distances,neighbors,begin,callback) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp:106:27: error: ‘n_neighbors’ not specified in enclosing ‘parallel’
  106 |     for (IndexType i=0; i<n_neighbors; i++)
      |                           ^~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp:53:9: error: enclosing ‘parallel’
   53 | #pragma omp parallel shared(shortest_distances,neighbors,begin,callback) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp: In function ‘tapkee::DenseMatrix tapkee::tapkee_internal::compute_shortest_distances_matrix(RandomAccessIterator, RandomAccessIterator, tapkee::tapkee_internal::Landmarks&, tapkee::tapkee_internal::Neighbors&, DistanceCallback) [with RandomAccessIterator = __gnu_cxx::__normal_iterator<int*, std::vector<int> >; DistanceCallback = pimpl_distance_callback<shogun::Distance>]’:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp:173:13: error: ‘N’ not specified in enclosing ‘parallel’
  173 |   bool* f = new bool[N];
      |             ^~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp:171:9: error: enclosing ‘parallel’
  171 | #pragma omp parallel shared(shortest_distances,begin,landmarks,neighbors,callback) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp:183:9: error: ‘N_landmarks’ not specified in enclosing ‘parallel’
  183 | #pragma omp for nowait
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp:171:9: error: enclosing ‘parallel’
  171 | #pragma omp parallel shared(shortest_distances,begin,landmarks,neighbors,callback) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp:224:27: error: ‘n_neighbors’ not specified in enclosing ‘parallel’
  224 |     for (IndexType i=0; i<n_neighbors; i++)
      |                           ^~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/isomap.hpp:171:9: error: enclosing ‘parallel’
  171 | #pragma omp parallel shared(shortest_distances,begin,landmarks,neighbors,callback) default(none)
      |         ^~~
In file included from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/methods.hpp:17,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/embed.hpp:11,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/tapkee.hpp:10,
                 from /home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/tapkee_shogun.cpp:20:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp: In function ‘tapkee::SparseWeightMatrix tapkee::tapkee_internal::hessian_weight_matrix(RandomAccessIterator, RandomAccessIterator, const Neighbors&, PairwiseCallback, tapkee::IndexType) [with RandomAccessIterator = __gnu_cxx::__normal_iterator<int*, std::vector<int> >; PairwiseCallback = pimpl_kernel_callback<shogun::Kernel>]’:
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:181:15: error: ‘k’ not specified in enclosing ‘parallel’
  181 |   DenseMatrix gram_matrix = DenseMatrix::Zero(k,k);
      |               ^~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:178:9: error: enclosing ‘parallel’
  178 | #pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:182:22: error: ‘target_dimension’ not specified in enclosing ‘parallel’
  182 |   DenseMatrix Yi(k,1+target_dimension+dp);
      |                      ^~~~~~~~~~~~~~~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:178:9: error: enclosing ‘parallel’
  178 | #pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
      |         ^~~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:182:39: error: ‘dp’ not specified in enclosing ‘parallel’
  182 |   DenseMatrix Yi(k,1+target_dimension+dp);
      |                                       ^~
/home/krakapwa/Documents/scratch_software/shogun/src/shogun/lib/tapkee/routines/locally_linear.hpp:178:9: error: enclosing ‘parallel’
  178 | #pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
      |         ^~~
make[2]: *** [src/shogun/CMakeFiles/libshogun.dir/build.make:5262: src/shogun/CMakeFiles/libshogun.dir/lib/tapkee/tapkee_shogun.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
^Cmake[2]: *** [src/shogun/CMakeFiles/libshogun.dir/build.make:5652: src/shogun/CMakeFiles/libshogun.dir/machine/gp/EPInferenceMethod.cpp.o] Interrupt
make[2]: *** [src/shogun/CMakeFiles/libshogun.dir/build.make:5678: src/shogun/CMakeFiles/libshogun.dir/machine/gp/FITCInferenceMethod.cpp.o] Interrupt
make[2]: *** [src/shogun/CMakeFiles/libshogun.dir/build.make:5730: src/shogun/CMakeFiles/libshogun.dir/machine/gp/KLCholeskyInferenceMethod.cpp.o] Interrupt
make[1]: *** [CMakeFiles/Makefile2:207: src/shogun/CMakeFiles/libshogun.dir/all] Interrupt
make: *** [Makefile:152: all] Interrupt
```

--- a/src/shogun/lib/tapkee/defines.hpp
+++ b/src/shogun/lib/tapkee/defines.hpp
@@ -37,6 +37,12 @@
 	#define COVERTREE_BASE 1.3
 #endif
 
+/* Compiler behavior workaround */
+#if defined(__GNUC__) && !defined(__clang__) && __GNUC__ < 9
+# define TAPKEE_NO_OMP_SHARED_CONSTANTS_
+#endif // defined(__GNUC__) && !defined(__clang__) && __GNUC__ < 9
+/* End of compiler behavior workaround */
+
 namespace tapkee
 {
 	//! Return result of the library - a pair of @ref DenseMatrix (embedding) and @ref ProjectingFunction

--- a/src/shogun/lib/tapkee/routines/diffusion_maps.hpp
+++ b/src/shogun/lib/tapkee/routines/diffusion_maps.hpp
@@ -45,7 +45,7 @@ DenseSymmetricMatrix compute_diffusion_matrix(RandomAccessIterator begin, Random
 	RESTRICT_ALLOC;
 
 	// compute gaussian kernel matrix
-#pragma omp parallel shared(diffusion_matrix,begin,callback) default(none)
+#pragma omp parallel shared(diffusion_matrix,begin,callback,width,n_vectors) default(none)
 	{
 		IndexType i_index_iter, j_index_iter;
 #pragma omp for nowait

--- a/src/shogun/lib/tapkee/routines/diffusion_maps.hpp
+++ b/src/shogun/lib/tapkee/routines/diffusion_maps.hpp
@@ -45,7 +45,11 @@ DenseSymmetricMatrix compute_diffusion_matrix(RandomAccessIterator begin, Random
 	RESTRICT_ALLOC;
 
 	// compute gaussian kernel matrix
-#pragma omp parallel shared(diffusion_matrix,begin,callback,width,n_vectors) default(none)
+#ifdef TAPKEE_NO_OMP_SHARED_CONSTANTS_
+# pragma omp parallel shared(diffusion_matrix,begin,callback) default(none)
+#else
+# pragma omp parallel shared(diffusion_matrix,begin,callback,width,n_vectors) default(none)
+#endif
 	{
 		IndexType i_index_iter, j_index_iter;
 #pragma omp for nowait

--- a/src/shogun/lib/tapkee/routines/isomap.hpp
+++ b/src/shogun/lib/tapkee/routines/isomap.hpp
@@ -50,7 +50,7 @@ DenseSymmetricMatrix compute_shortest_distances_matrix(RandomAccessIterator begi
 
 	DenseSymmetricMatrix shortest_distances(N,N);
 
-#pragma omp parallel shared(shortest_distances,neighbors,begin,callback) default(none)
+#pragma omp parallel shared(shortest_distances,neighbors,begin,callback,N,n_neighbors) default(none)
 	{
 		bool* f = new bool[N];
 		bool* s = new bool[N];
@@ -168,7 +168,7 @@ DenseMatrix compute_shortest_distances_matrix(RandomAccessIterator begin, Random
 
 	DenseMatrix shortest_distances(landmarks.size(),N);
 
-#pragma omp parallel shared(shortest_distances,begin,landmarks,neighbors,callback) default(none)
+#pragma omp parallel shared(shortest_distances,begin,landmarks,neighbors,callback,N,N_landmarks,n_neighbors) default(none)
 	{
 		bool* f = new bool[N];
 		bool* s = new bool[N];

--- a/src/shogun/lib/tapkee/routines/isomap.hpp
+++ b/src/shogun/lib/tapkee/routines/isomap.hpp
@@ -50,7 +50,11 @@ DenseSymmetricMatrix compute_shortest_distances_matrix(RandomAccessIterator begi
 
 	DenseSymmetricMatrix shortest_distances(N,N);
 
-#pragma omp parallel shared(shortest_distances,neighbors,begin,callback,N,n_neighbors) default(none)
+#ifdef TAPKEE_NO_OMP_SHARED_CONSTANTS_
+# pragma omp parallel shared(shortest_distances,neighbors,begin,callback) default(none)
+#else
+# pragma omp parallel shared(shortest_distances,neighbors,begin,callback,N,n_neighbors) default(none)
+#endif
 	{
 		bool* f = new bool[N];
 		bool* s = new bool[N];
@@ -168,7 +172,11 @@ DenseMatrix compute_shortest_distances_matrix(RandomAccessIterator begin, Random
 
 	DenseMatrix shortest_distances(landmarks.size(),N);
 
-#pragma omp parallel shared(shortest_distances,begin,landmarks,neighbors,callback,N,N_landmarks,n_neighbors) default(none)
+#ifdef TAPKEE_NO_OMP_SHARED_CONSTANTS_
+# pragma omp parallel shared(shortest_distances,begin,landmarks,neighbors,callback) default(none)
+#else
+# pragma omp parallel shared(shortest_distances,begin,landmarks,neighbors,callback,N,N_landmarks,n_neighbors) default(none)
+#endif
 	{
 		bool* f = new bool[N];
 		bool* s = new bool[N];

--- a/src/shogun/lib/tapkee/routines/locally_linear.hpp
+++ b/src/shogun/lib/tapkee/routines/locally_linear.hpp
@@ -32,7 +32,7 @@ SparseWeightMatrix tangent_weight_matrix(RandomAccessIterator begin, RandomAcces
 	SparseTriplets sparse_triplets;
 	sparse_triplets.reserve((k*k+2*k+1)*(end-begin));
 
-#pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
+#pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets,k,target_dimension,shift) default(none)
 	{
 		IndexType index_iter;
 		DenseMatrix gram_matrix = DenseMatrix::Zero(k,k);
@@ -102,7 +102,7 @@ SparseWeightMatrix linear_weight_matrix(const RandomAccessIterator& begin, const
 	SparseTriplets sparse_triplets;
 	sparse_triplets.reserve((k*k+2*k+1)*(end-begin));
 
-#pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
+#pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets,k,shift,trace_shift) default(none)
 	{
 		IndexType index_iter;
 		DenseMatrix gram_matrix = DenseMatrix::Zero(k,k);
@@ -175,7 +175,7 @@ SparseWeightMatrix hessian_weight_matrix(RandomAccessIterator begin, RandomAcces
 
 	const IndexType dp = target_dimension*(target_dimension+1)/2;
 
-#pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
+#pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets,k,target_dimension,dp) default(none)
 	{
 		IndexType index_iter;
 		DenseMatrix gram_matrix = DenseMatrix::Zero(k,k);

--- a/src/shogun/lib/tapkee/routines/locally_linear.hpp
+++ b/src/shogun/lib/tapkee/routines/locally_linear.hpp
@@ -32,7 +32,11 @@ SparseWeightMatrix tangent_weight_matrix(RandomAccessIterator begin, RandomAcces
 	SparseTriplets sparse_triplets;
 	sparse_triplets.reserve((k*k+2*k+1)*(end-begin));
 
-#pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets,k,target_dimension,shift) default(none)
+#ifdef TAPKEE_NO_OMP_SHARED_CONSTANTS_
+# pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
+#else
+# pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets,k,target_dimension,shift) default(none)
+#endif
 	{
 		IndexType index_iter;
 		DenseMatrix gram_matrix = DenseMatrix::Zero(k,k);
@@ -102,7 +106,11 @@ SparseWeightMatrix linear_weight_matrix(const RandomAccessIterator& begin, const
 	SparseTriplets sparse_triplets;
 	sparse_triplets.reserve((k*k+2*k+1)*(end-begin));
 
-#pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets,k,shift,trace_shift) default(none)
+#ifdef TAPKEE_NO_OMP_SHARED_CONSTANTS_
+# pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
+#else
+# pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets,k,shift,trace_shift) default(none)
+#endif
 	{
 		IndexType index_iter;
 		DenseMatrix gram_matrix = DenseMatrix::Zero(k,k);
@@ -175,7 +183,11 @@ SparseWeightMatrix hessian_weight_matrix(RandomAccessIterator begin, RandomAcces
 
 	const IndexType dp = target_dimension*(target_dimension+1)/2;
 
-#pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets,k,target_dimension,dp) default(none)
+#ifdef TAPKEE_NO_OMP_SHARED_CONSTANTS_
+# pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets) default(none)
+#else
+# pragma omp parallel shared(begin,end,neighbors,callback,sparse_triplets,k,target_dimension,dp) default(none)
+#endif
 	{
 		IndexType index_iter;
 		DenseMatrix gram_matrix = DenseMatrix::Zero(k,k);

--- a/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp
+++ b/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp
@@ -37,7 +37,11 @@ DenseSymmetricMatrix compute_distance_matrix(RandomAccessIterator begin, RandomA
 	const IndexType n_landmarks = landmarks.size();
 	DenseSymmetricMatrix distance_matrix(n_landmarks,n_landmarks);
 
-#pragma omp parallel shared(begin,landmarks,distance_matrix,callback,n_landmarks) default(none)
+#ifdef TAPKEE_NO_OMP_SHARED_CONSTANTS_
+# pragma omp parallel shared(begin,landmarks,distance_matrix,callback) default(none)
+#else
+# pragma omp parallel shared(begin,landmarks,distance_matrix,callback,n_landmarks) default(none)
+#endif
 	{
 		IndexType i_index_iter,j_index_iter;
 #pragma omp for nowait
@@ -79,8 +83,13 @@ DenseMatrix triangulate(RandomAccessIterator begin, RandomAccessIterator end, Pa
 	for (IndexType i=0; i<target_dimension; ++i)
 		landmarks_embedding.first.col(i).array() /= landmarks_embedding.second(i);
 
-#pragma omp parallel shared(begin,end,to_process,distance_callback,landmarks, \
+#ifdef TAPKEE_NO_OMP_SHARED_CONSTANTS_
+# pragma omp parallel shared(begin,end,to_process,distance_callback,landmarks, \
+		landmarks_embedding,landmark_distances_squared,embedding) default(none)
+#else
+# pragma omp parallel shared(begin,end,to_process,distance_callback,landmarks, \
 		landmarks_embedding,landmark_distances_squared,embedding,n_landmarks,n_vectors) default(none)
+#endif
 	{
 		DenseVector distances_to_landmarks(n_landmarks);
 		IndexType index_iter;
@@ -116,7 +125,11 @@ DenseSymmetricMatrix compute_distance_matrix(RandomAccessIterator begin, RandomA
 	const IndexType n_vectors = end-begin;
 	DenseSymmetricMatrix distance_matrix(n_vectors,n_vectors);
 
-#pragma omp parallel shared(begin,distance_matrix,callback,n_vectors) default(none)
+#ifdef TAPKEE_NO_OMP_SHARED_CONSTANTS_
+# pragma omp parallel shared(begin,distance_matrix,callback) default(none)
+#else
+# pragma omp parallel shared(begin,distance_matrix,callback,n_vectors) default(none)
+#endif
 	{
 		IndexType i_index_iter,j_index_iter;
 #pragma omp for nowait

--- a/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp
+++ b/src/shogun/lib/tapkee/routines/multidimensional_scaling.hpp
@@ -37,7 +37,7 @@ DenseSymmetricMatrix compute_distance_matrix(RandomAccessIterator begin, RandomA
 	const IndexType n_landmarks = landmarks.size();
 	DenseSymmetricMatrix distance_matrix(n_landmarks,n_landmarks);
 
-#pragma omp parallel shared(begin,landmarks,distance_matrix,callback) default(none)
+#pragma omp parallel shared(begin,landmarks,distance_matrix,callback,n_landmarks) default(none)
 	{
 		IndexType i_index_iter,j_index_iter;
 #pragma omp for nowait
@@ -80,7 +80,7 @@ DenseMatrix triangulate(RandomAccessIterator begin, RandomAccessIterator end, Pa
 		landmarks_embedding.first.col(i).array() /= landmarks_embedding.second(i);
 
 #pragma omp parallel shared(begin,end,to_process,distance_callback,landmarks, \
-		landmarks_embedding,landmark_distances_squared,embedding) default(none)
+		landmarks_embedding,landmark_distances_squared,embedding,n_landmarks,n_vectors) default(none)
 	{
 		DenseVector distances_to_landmarks(n_landmarks);
 		IndexType index_iter;
@@ -116,7 +116,7 @@ DenseSymmetricMatrix compute_distance_matrix(RandomAccessIterator begin, RandomA
 	const IndexType n_vectors = end-begin;
 	DenseSymmetricMatrix distance_matrix(n_vectors,n_vectors);
 
-#pragma omp parallel shared(begin,distance_matrix,callback) default(none)
+#pragma omp parallel shared(begin,distance_matrix,callback,n_vectors) default(none)
 	{
 		IndexType i_index_iter,j_index_iter;
 #pragma omp for nowait

